### DISCUSSION TITLE
help ili9488

### DISCUSSION THREAD

**Discussion by Roman26rus** [Bodmer/TFT_eSPI#1091] [Q&A - Touch]
hello. help please I use esp32 I connect the LCD via the parallel interface, I connect the touchscreen via SPI, how to use the touch screen via SPI code the mode of inclusion is parallel ? The ILI9488 LCD controller .touch screen XPT2046 .the problem is that the LCD only works on a parallel 8bit interface ,and the touch is connected to SPI,can you somehow make it work with the built-in library of the touch screen?

**Comment by Pablo2048** [comment]
https://github.com/Bodmer/TFT_eSPI#touch-controller-support - read the first sentence in the paragraph...

**Comment by Roman26rus** [comment]
![800px-3 95-Mega2560-002](https://user-images.githubusercontent.com/42438844/111869362-212faa80-8990-11eb-9daf-f00957711fda.jpg)

**Comment by Roman26rus** [comment]
well, in fact, you can make it work, but I tried the library <XPT2046_Touchscreen.h> it works in conjunction with TFT_eSPI
just I would like to just enable SPI reading of the touch screen in the display connection mode via a parallel bus

**Comment by Pablo2048** [comment]
There is clearly written "The XPT2046 touch screen controller is supported for SPI based displays only." There was many discussions about this (did you search the closed issues?), for example here https://github.com/Bodmer/TFT_eSPI/issues/979#issuecomment-772109636 , https://github.com/Bodmer/TFT_eSPI/discussions/895 ...
EDIT: ... and here https://github.com/Bodmer/TFT_eSPI/issues/947

**Reply by Roman26rus** [reply]
Я до сих пор не понял, как использовать его в качестве примера . есть ли у вас какой-нибудь пример того, как сделать кнопку ? нажав на которую что-то произошло, я не могу понять, как использовать библиотеку touch

**Comment by Bodmer** [comment]
As noted by others above. #895 gives example code.

**Reply by Roman26rus** [reply]
I still haven't figured out how to use it as an example . do you have any example on how to make a button ? by pressing which something happened, I can not understand how to use the library touch

**Reply by Roman26rus** [reply]
and one more question ,could you make your library support the parallel connection mode and can also work with touch ? could you add so that the examples that are in the library work ? or is it difficult? so as not to put separate libraries on the touch and use the built in and the examples all worked without alteration

**Comment by Bodmer** [comment]
The aim in the longer term is to deprecate the touch support built into TFT_eSPI as there are better touch libraries available that are good.

Try [this library](https://github.com/achillhasler/TFT_eTouch), it supports the XPT2046 touch controller with TFT_eSPI parallel screens.  Probably only the "calibrate" and "raw" sketches will work. The examples assume you are using pin 3 (corrected) as the touch CS line.

Once you have the sketches working post back.

**Reply by Bodmer** [reply]
Correction the example [assumes pin 3 for touch CS](https://github.com/achillhasler/TFT_eTouch/blob/master/TFT_eTouchUser.h#L24).

**Reply by Roman26rus** [reply]
can you show an example of how to set the coordinates for the touch, using the example of a normal button ? I don't understand how to use this library ,it works in calibration mode ,but I don't understand how to use it in a project using the example of a regular button .can you show me ? please

**Reply by Bodmer** [reply]
First write a sketch that draws a the "button" on the screen without trying to use touch. Post your sketch here.

You can use [this example sketch](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Generic/On_Off_Button/On_Off_Button.ino) as a starting point.

**Reply by Roman26rus** [reply]
#include <XPT2046_Touchscreen.h>
#include <SPI.h>
#include <TFT_eSPI.h>      // Hardware-specific library
TFT_eSPI tft = TFT_eSPI(); // Invoke custom library
#define CS_PIN  2
#define TIRQ_PIN  35
XPT2046_Touchscreen ts(CS_PIN, TIRQ_PIN); 

bool ON_OFF_1 = false; // флаг, указывающий состояние кнопки
bool canTouchThis_1 = true; // флаг создан для того, чтобы избежать переключения состояний при удержании кнопки

bool ON_OFF_2 = false; // флаг, указывающий состояние кнопки
bool canTouchThis_2 = true; // флаг создан для того, чтобы избежать переключения состояний при удержании кнопки

float lcd_x = 0; //удерживает положение касания X
float lcd_y = 0; //удерживает положение касания Y
float lcd_x_2 = 0; //удерживает положение касания X
float lcd_y_2 = 0; //удерживает положение касания Y

#define TFT_W 480
#define TFT_H 320

void setup() {
 Serial.begin(115200);
 tft.init();
 ts.begin();
 tft.setRotation(1);
 ts.setRotation(3);
 #define TOUCH_DEFAULT_CALIBRATION { 332, 3722, 3847, 248, 0 }
  //while (!Serial && (millis() <= 1000));
//  drawScreen_2("2", TFT_WHITE, TFT_DARKGREY);
 drawScreen_1("1", TFT_WHITE, TFT_DARKGREY);
  
}

void loop() {
 
  if(ts.tirqTouched()){
    if (ts.touched()) {
 TS_Point p = ts.getPoint();
      lcd_x = p.x/(4096/TFT_W); // вычисляет точку на основе ширины экрана
      lcd_y = p.y/(4096/TFT_H);// вычисляет точку на основе высоты экрана
      if((lcd_x > 80 and lcd_x < 239) and (lcd_y > 70 and lcd_y < 169) and  canTouchThis_1){ // if((lcd_x_2 > 10 and lcd_x_2 < 150) and (lcd_y_2 > 70 and lcd_y_2 < 50) and  canTouchThis_2){
////////////////////////////////////////////////////////////////////      
      if(ON_OFF_1){
      drawScreen_1("[OFF]", TFT_WHITE, TFT_RED);
      }else{
      drawScreen_1("[ON]", TFT_BLACK,TFT_GREEN);
      }
       ON_OFF_1 = !ON_OFF_1; //switch states
 canTouchThis_1 = false;
  }

  }
}else{canTouchThis_1 = true;}
}


void drawScreen_1(String s_1, uint16_t color, uint16_t bg_1){
    tft.fillScreen(TFT_BLACK);    
    tft.fillRoundRect(80,70, 239-80,169-70, 1,bg_1); //прямоугольник. расположение на диспл (X горизонт , Y вертикаль),H-ширина,W-высота,закругление,цвет
    //tft.setTextColor(ILI9341_WHITE,TFT_DARK_GREY);
    tft.setTextColor(color);
     tft.drawCentreString(s_1, 160, 112,4);//рисует строку (тип String) с центром в позиции X
    
  }

**Reply by Roman26rus** [reply]
Here is such I try to do code, but not korektno works sensor, the code that you gave as example not works error touch, with you there is opportunity communicate in Telegram ?or WhatsApp?

**Reply by Bodmer** [reply]
I will have a look and see what I can do to help. For various reasons this is the only communication path.

**Comment by TangerineDreamer** [comment]
All is in the Keypad example,first you configure the touch calibrate and then you can press buttons
and see the result on the screen.
Maybe you will have to make the SPIFFS cleared before the calibrate,eg: if you change the rotate of the sketch.
